File: leetcode/Binary_Search/0069_Sqrt_x.py
class Solution(object):
    def mySqrt(self, x):
        """
        :type x: int
        :rtype: int
        """
        l , r = 0, 46340 # sqrt(2^31 - 1) = 46340.95
        while l <= r:
            mid = (l + r) // 2
            if mid * mid == x:
                return mid
            elif mid * mid < x:
                l = mid + 1
            else:
                r = mid - 1

        return r

File: leetcode/Binary_Search/0069_Sqrt_x.cpp
// Newton's method
class Solution {
public:
	int mySqrt(int x) {
		long long root = x;
		while (root * root > x) {
			root = (root + x / root) / 2;
		}
		return root;
	}
};

// binary search
class Solution {
public:
	int mySqrt(int x) {
		int l = 0, r = 46340;
		while(l <= r) {
			int mid = (l + r) / 2;
			if (mid * mid == x)
				return mid;
			else if (mid * mid < x)
				l = mid + 1;
			else
				r = mid - 1;
		}
		return r;
	}
};
